src/coreclr/vm/exceptionhandling.cpp

```diff
@@ -648,28 +665,33 @@ ProcessCLRException(IN     PEXCEPTION_RECORD   pExceptionRecord,
             EEPOLICY_HANDLE_FATAL_ERROR(pExceptionRecord->ExceptionCode);
         }
 
+#ifdef TARGET_X86
+        CallRtlUnwind((PEXCEPTION_REGISTRATION_RECORD)pEstablisherFrame, NULL, pExceptionRecord, 0);
```

Comment:
It is strange that after this returns, we continue running the 2nd pass stuff. I would think that the RtlUnwind would have already called the ProcessCLRException for the current frame on the 2nd pass, so it seems to me we are doing the work below twice if the RtlUnwind returned. I actually wonder what is the case when it returns, because it should call the DispatchManagedException on the 2nd pass and that one would never return.

Reply:
`RtlUnwind` on x86 behaves quite differently than on other platforms. It essentially unwinds everything up to the frame that you passed in the first parameter and then you are responsible for running the actual unwinding for your current frame.

Reply:
But it still calls the ProcessCLRException while walking the stack, that was my point. So once it returns, it would have called the ProcessCLRException even for the frame for which we've invoked the RtlUnwind.

Reply:
I am pretty sure it stops at the frame passed in the first parameter without executing its personality routine.

Reply:
For legal reasons I cannot point to / read WRK source code, so the next best is Wine implementation:
https://github.com/wine-mirror/wine/blob/e956c4ec71dd0f41090df3863e6f937963b2b7d9/dlls/ntdll/signal_i386.c#L321-L402

It captures the context at the beginning of `RtlUnwind` (patches ESP, EAX, so it essentially points to the return address after `RtlUnwind` with specified return value), does the unwinding loop until `frame != pEndFrame`, and then runs `NtContinue` (aka `RtlRestoreContext`) to exit out as if normally returning from the function.

Reply:
That's strange, I've looked at the original Windows implementation and there is a comment saying that it calls exception handlers as they are encountered during the unwinding. And even in the Wine one you've pointed out, it is iterating over the records and calling the handlers, so I guess I must be missing something. I have to admit that the x86 SEH / .NET EH is something I've managed to mostly avoid over the years, so it is possible that I am missing some fundamental detail.

Reply:
Ah, ok, I was expecting it to call it for the one that called the RtlUnwind too. That's what happens on other architectures. 
I've tried to re-read the code of the RtlUnwind and it seems that we would get the same behavior if we passed NULL instead of the establisherFrame to the RtlUnwind. In that case, the RtlUnwind would never return and call the ProcessCLRException for the handler that called the RtlUnwind too. That one would call the DispatchManagedException, so it would not matter that we didn't pass in any target frame.
But let's leave your PR as is and we can try to play with this after.

Reply:
(IIRC `NULL` would unwind everything including handlers below on the stack. I didn't study that behavior in detail though.)